src/Common/src/System/Security/Cryptography/Asn1V2.Serializer.cs

```diff
@@ -0,0 +1,1431 @@
+                        SR.Cryptography_AsnSerializer_Choice_NoChoiceWasMade,
+                        typeT.FullName));
+            }
+
+            Serializer serializer = GetSerializer(relevantField.FieldType, relevantField);
+            serializer(relevantObject, writer);
+        }
+
+        private static object DeserializeChoice(AsnReader reader, Type typeT)
+        {
+            var lookup = new Dictionary<(TagClass, int), LinkedList<FieldInfo>>();
+            LinkedList<FieldInfo> fields = new LinkedList<FieldInfo>();
+            PopulateChoiceLookup(lookup, typeT, fields);
+
+            Asn1Tag next = reader.PeekTag();
+
+            if (next == Asn1Tag.Null)
+            {
+                ChoiceAttribute choiceAttr = GetChoiceAttribute(typeT);
+
+                if (choiceAttr.AllowNull)
+                {
+                    reader.ReadNull();
+                    return null;
+                }
+
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            var key = (next.TagClass, next.TagValue);
+
+            if (lookup.TryGetValue(key, out LinkedList<FieldInfo> fieldInfos))
+            {
+                LinkedListNode<FieldInfo> currentNode = fieldInfos.Last;
+                FieldInfo currentField = currentNode.Value;
+                object currentObject = Activator.CreateInstance(currentField.DeclaringType);
+                Deserializer deserializer = GetDeserializer(currentField.FieldType, currentField);
+                object deserialized = deserializer(reader);
+                currentField.SetValue(currentObject, deserialized);
+
+                while (currentNode.Previous != null)
+                {
+                    currentNode = currentNode.Previous;
+                    currentField = currentNode.Value;
+
+                    object nextObject = Activator.CreateInstance(currentField.DeclaringType);
+                    currentField.SetValue(nextObject, currentObject);
+
+                    currentObject = nextObject;
+                }
+
+                return currentObject;
+            }
+
+            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+        }
+
+        private static void SerializeCustomType(Type typeT, object value, AsnWriter writer, Asn1Tag tag)
+        {
+            writer.PushSequence(tag);
+
+            foreach (FieldInfo fieldInfo in typeT.GetFields(FieldFlags))
+            {
+                Serializer serializer = GetSerializer(fieldInfo.FieldType, fieldInfo);
+                serializer(fieldInfo.GetValue(value), writer);
+            }
+
+            writer.PopSequence(tag);
+        }
+
+        private static object DeserializeCustomType(AsnReader reader, Type typeT)
+        {
+            object target = Activator.CreateInstance(typeT);
+
+            AsnReader sequence = reader.ReadSequence();
+
+            foreach (FieldInfo fieldInfo in typeT.GetFields(FieldFlags))
```

Comment:
We need to investigate whether GetFields is guaranteed stable ordering between CLR families (Core, Desktop, ...) and versions (1.1, 2.0, 4.0, ...). And also add a comment that we rely on this behavior.